sdks/python/src/opik/integrations/adk/opik_tracer.py

```diff
@@ -53,49 +137,58 @@ def before_agent_callback(
             thread_id = str(uuid.uuid4())
             callback_context.state["opik_thread_id"] = thread_id
 
-        # Should we create a trace here as we don't have an input?
-
         trace_metadata = self.metadata.copy()
         trace_metadata["adk_invocation_id"] = callback_context.invocation_id
 
-        user_input = adk_decorators.convert_adk_base_models(
+        user_input = adk_helpers.convert_adk_base_model_to_dict(
             callback_context.user_content
         )
+        name = self.name or callback_context.agent_name
 
-        self.trace_data = trace.TraceData(
-            id=helpers.generate_id(),
-            start_time=datetime_helpers.local_timestamp(),
-            name=self.name or callback_context.agent_name,
-            input=user_input,
-            metadata=trace_metadata,
-            tags=self.tags,
-            project_name=self.project_name,
-            thread_id=thread_id,
-        )
-
-        context_storage.set_trace_data(self.trace_data)
+        if (current_span_data := self._context_storage.top_span_data()) is not None:
+            self._attach_span_to_existing_span(
+                current_span_data=current_span_data,
+                name=name,
+                input=user_input,
+                type="general",
+                metadata=self.metadata,
+            )
+        elif (current_trace_data := self._context_storage.get_trace_data()) is not None:
+            self._attach_span_to_existing_trace(
+                current_trace_data=current_trace_data,
+                name=name,
+                input=user_input,
+                type="general",
+                metadata=self.metadata,
+            )
+        else:
+            new_trace_data = trace.TraceData(
+                name=name,
+                input=user_input,
+                metadata=self.metadata,
+                project_name=self.project_name,
+                thread_id=thread_id,
+            )
+            self._start_trace(new_trace_data)
 
     def after_agent_callback(
         self, callback_context: CallbackContext, *args: Any, **kwargs: Any
     ) -> None:
-        assert context_storage.get_trace_data() is self.trace_data
-        assert self.trace_data is not None
-
-        context_storage.pop_trace_data()
-
         output = self._last_model_output
 
         # remove the custom metadata with opik usage we added
-        if output is not None:
-            llm_response_wrapper.pop_llm_usage_data(**output)
-
-        self.trace_data.update(output=output).init_end_time()
-
-        self._client.trace(**self.trace_data.__dict__)
+        if output is not None and "custom_metadata" in output:
+            output.pop("custom_metadata")
```

Comment:
Perhaps we should avoid removing the entire `custom_metadata` here, as some sections might have been set by other components, not just by us.

Reply:
Good catch, I forgot it's not our field